### PR TITLE
Use QS Groups as QS Assets Owners

### PR DESCRIPTION
**Pull request by iakov-aws**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cid/helpers/quicksight/__init__.py [hunk 1]
```diff
@@ -233,12 +241,83 @@ def _recoursive_add_view(view):
...
+        if self._principal_arn:
+            return self._principal_arn
+
+        if self.group: self._principal_arn = self.group.get('Arn')
+        elif self.user: self._principal_arn = self.user.get('Arn')
+
+        if self._principal_arn:
+            return self._principal_arn
+
+        quicksight_owner = get_parameter('quicksight-owner-choice',
```
Comment (iakov-aws) [inline cid/helpers/quicksight/__init__.py:271]: Add a comment : Not a real parameter. Not supposed to be used by CLI users

File: cid/helpers/quicksight/__init__.py [hunk 2]
```diff
@@ -233,12 +241,83 @@ def _recoursive_add_view(view):
...
+            ).get('Group')
+        return group
+
+
+    def get_principal_arn(self):
+        if self._principal_arn:
+            return self._principal_arn
+
+        if self.group: self._principal_arn = self.group.get('Arn')
+        elif self.user: self._principal_arn = self.user.get('Arn')
```
Comment (iakov-aws) [inline cid/helpers/quicksight/__init__.py:266]: TODO: Fail if both

**Comment by iakov-aws** [comment]
@darken99 @yprikhodko ready for merge

File: cid/helpers/quicksight/__init__.py [hunk 3]
```diff
@@ -234,12 +243,88 @@ def _recoursive_add_view(view):
             logger.info(f'"{dashboard.name}" ({dashboardId}) discover complete')
             return dashboard
 
+    def ensure_exist_group(self, groupname='cid-owners', description='Created by Cloud Intelligence Dashboards'):
```
Comment (darken99) [inline cid/helpers/quicksight/__init__.py:246]: Please rename to follow common naming

File: cid/helpers/quicksight/__init__.py [hunk 4]
```diff
@@ -234,12 +243,88 @@ def _recoursive_add_view(view):
...
+                Namespace='default',
+                GroupName=groupname
+            ).get('Group')
+        except self.client.exceptions.ResourceNotFoundException:
+            group = self.identityClient.create_group(
+                AwsAccountId=self.account_id,
+                GroupName=groupname,
+                Namespace='default',
+                description=description,
+            ).get('Group')
```
Comment (darken99) [inline cid/helpers/quicksight/__init__.py:259]: We need to handle AccessDenied exception too

File: cid/helpers/quicksight/__init__.py [hunk 5]
```diff
@@ -764,12 +849,37 @@ def describe_user(self, username: str) -> dict:
...
+                'Namespace': 'default'
+            })
+            logger.debug('group = ',json.dumps(result))
+            return result.get('Group')
+        except self.client.exceptions.ResourceNotFoundException:
+            logger.info(f'QuickSight group {groupname} not found.')
+            return None
+        except self.client.exceptions.AccessDeniedException: # Try to overcome AccessDeniedException
+            #FIXME: paginator is not available for list_groups
+            groups = self.identityClient.list_groups(AwsAccountId=self.account_id, Namespace='default').get('GroupList')
```
Comment (darken99) [inline cid/helpers/quicksight/__init__.py:867]: This can throw an exception too

File: cid/helpers/quicksight/__init__.py (hunk 4 above)
Comment (iakov-aws) [reply cid/helpers/quicksight/__init__.py:259]: fixed

File: cid/helpers/quicksight/__init__.py (hunk 3 above)
Comment (iakov-aws) [reply cid/helpers/quicksight/__init__.py]: fixed

File: cid/helpers/quicksight/__init__.py (hunk 5 above)
Comment (iakov-aws) [reply cid/helpers/quicksight/__init__.py]: i guess we do not need this workaround here. simplified.

**Review by darken99 — APPROVED** [review]

**Pull request merged**